README.md
This package contains helpers for serving blogs created with [Hyvor Blogs](https://hyvor.com/blogs) on web applications such as Next.js. It can run on any framework that supports Web APIs.

## Tutorials

Here are some tutorials on how to use this package with different frameworks.

* [Next.js](https://hyvor.com/blog/nextjs-blog)
* [Cloudflare Workers](https://hyvor.com/blog/cloudflare-workers-blog)

## Installation

```bash
npm install @hyvor/hyvor-blogs-serve-web
```

## Usage

First, create a new instance of `Blog`

```ts
import { Blog } from '@hyvor/hyvor-blogs-serve-web';

const blog = new Blog({

    /**
     * The subdomain of your blog.
     * Console -> Settings -> Hosting
     */
    subdomain: 'my-subdomain',

    /**
     * The Delivery API key of your blog.
     * Console -> Settings -> API Keys
     */
    deliveryApiKey: 'my-delivery-api-key',

    /**
     * @optional
     * The webhook secret key of your blog.
     * Console -> Settings -> Webhooks
     */
    webhookSecret: 'my-webhook-secret',

    /**
     * An object that has the following methods:
     * get: (key: string) => Promise<any>
     * set: (key: string, value: any) => Promise<any>
     * delete: (key: string) => Promise<any>
     */
    cache: myCache,

})
```

### Caching

Setting up a cache is highly recommended to reduce the number of requests to our Delivery API and to improve the performance of your blog. We recommend using [Keyv](https://keyvhq.js.org/) as the caching layer. If no cache is provided, Keyv in-memory cache is used, which does not give any performance benefits.

Here is an example using Redis as the cache store.

```bash
npm install @keyvhq/redis --save
```

```ts
import Keyv from '@keyvhq/core'
import KeyvRedis from '@keyvhq/redis'

const blog = new Blog({
    // ...
    cache: new Keyv({store: new KeyvRedis('redis://user:pass@localhost:6379')})
})
```

See [Keyv's documentation](https://keyvhq.js.org/#/?id=all-the-adapters) for all the available adapters.

#### Using a custom cache

If you want to use a custom cache, you can pass an object that has the following methods:

```ts
new Blog({
    cache: {
        get: async (key: string) => {
            // return the value of the key
        },
        set: async (key: string, value: any) => {
            // set the value of the key
        },
        delete: async (key: string) => {
            // delete the key
        }
    }
});
```

### Handling blog requests

Use the `handleRequest` method to handle blog requests. 


```ts
const response = await blog.handleRequest(path)
```

* It accepts one parameter: `path`, which is the path of the request. For example, if your blog root is `/blog`, for a request to `/blog/hello-world`, the `path` parameter should be `/hello-world`. This is used to call our [Delivery API](https://blogs.hyvor.com/docs/api-delivery).
* It returns a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) object.

### Handling webhooks

We use webhooks to clear cache of your blog when data is updated in your blog. You can handle webhooks using the `handleWebhook` method.

```ts
await blog.handleWebhook(data, signature)
```

* It accepts two parameters:
    * `data`: The request body as an object.
    * `signature`: The `X-Signature` header of the request.


Here is an example using the `Request` object.

```ts
const body = await request.json();
const signature = request.headers.get('X-Signature') || '';

await blog.handleWebhook(body, signature);
```

See our [Webhooks](https://blogs.hyvor.com/docs/webhooks) docs for more information.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## 1.1.3 - 2023-07-22

## 1.1.2 - 2023-07-22

## 1.1.1 - 2023-07-22

## 1.1.0 - 2023-07-22
### Added
- Changelog

## 1.0.5 - 2023-09-12

## 1.0.4 - 2023-09-12

## 1.0.3 - 2023-07-25

## 1.0.2 - 2023-07-25

## 1.0.1 - 2023-07-25


tests/blog.test.ts
import { describe, expect, it, vi } from "vitest";
import { Blog } from "../src";
import { BlogOptions, ResponseObject } from "../src/types";
import Keyv from "@keyvhq/core";

export function getBlog(options: Partial<BlogOptions> = {}) {

    return new Blog({
        subdomain: 'test',
        deliveryApiKey: 'delivery-api-key',
        cache: new Keyv({ store: new Map() }), // in-memory cache
        ...options
    })

}


it('returns blog response success', async () => {

    const deliveryResponseTime = Math.floor(Date.now() / 1000);
    const deliveryResponse : ResponseObject = {
        type: 'file',
        file_type: 'template',
        at: deliveryResponseTime,
        cache: true,
        cache_control: 'no-cache',
        status: 200,
        content: btoa('Hello World'),
        mime_type: 'text/html',
    }

    const mock = vi.fn().mockResolvedValue({
        ok: true,
        json: () => new Promise(resolve =>  resolve(deliveryResponse))
    });
    global.fetch = mock;

    const blog = getBlog();
    const response = await blog.handleBlogRequest('');

    expect(response.status).toBe(200);
    expect(response.headers.get('Cache-Control')).toBe('no-cache');

    expect(await response.text()).toBe('Hello World');

    expect(mock).toHaveBeenCalledOnce();
    expect(mock).toHaveBeenCalledWith(
        'https://blogs.hyvor.com/api/delivery/v0/test' +
        '?path=/' +
        '&api_key=delivery-api-key',
        { cache: 'no-store' }
    )

    const cached = await blog.cacheService.get('/');
    expect(cached).toEqual(deliveryResponse);

});

it('returns blog redirect response with path', async () => {
   
    const deliveryResponseTime = Math.floor(Date.now() / 1000);
    const deliveryResponse : ResponseObject = {
        type: 'redirect',
        at: deliveryResponseTime,
        cache: true,
        status: 301,
        to: 'https://example.com',
    }

    const mock = vi.fn().mockResolvedValue({
        ok: true,
        json: () => new Promise(resolve =>  resolve(deliveryResponse))
    });
    global.fetch = mock;

    const blog = getBlog();
    const response = await blog.handleBlogRequest('/about');

    expect(response.status).toBe(301);
    expect(response.headers.get('Location')).toBe('https://example.com/');

    expect(mock).toHaveBeenCalledOnce();
    expect(mock).toHaveBeenCalledWith(
        'https://blogs.hyvor.com/api/delivery/v0/test' +
        '?path=/about' +
        '&api_key=delivery-api-key',
        { cache: 'no-store' }
    )

    const cached = await blog.cacheService.get('/about');
    expect(cached).toEqual(deliveryResponse);
    
});

it('handles server error', async () => {

    const mock = vi.fn().mockResolvedValue({
        ok: false,
    });
    global.fetch = mock;

    const blog = getBlog();
    const response = await blog.handleBlogRequest('/bad');

    expect(response.status).toBe(500);

    const cached = await blog.cacheService.get('/about');
    expect(cached).toBeNull();

});


describe('experimental', function() {

    it('no cache in fetch', async () => {

        
        const mock = vi.fn().mockResolvedValue({
            ok: false,
        });
        global.fetch = mock;

        const blog = getBlog({
            experimental: {
                dontSetCacheInFetch: true
            }
        });

        await blog.handleBlogRequest('/');
        expect(mock).toHaveBeenCalledOnce();
        expect(mock).toHaveBeenCalledWith(
            'https://blogs.hyvor.com/api/delivery/v0/test' +
            '?path=/' +
            '&api_key=delivery-api-key',
            {}
        );

    });

});

tests/webhook.test.ts
import { expect, it } from "vitest";
import { getBlog } from "./blog.test";
import { ResponseObject } from "../src/types";

it('requires a valid object', async () => {
    expect(async () => await getBlog().handleWebhookRequest('invalid', 'sig'))
        .rejects
        .toThrowError('Invalid data type');
})

it('should match the subdomain', async () => {
    expect(async() => await getBlog().handleWebhookRequest({subdomain: 'test2'}, 'sig'))
        .rejects    
        .toThrowError('Subdomain mismatch');
});

it('should match the signature', async () => {
    expect(async() => await getBlog({
        webhookSecret: 'secret'
    }).handleWebhookRequest({subdomain: 'test'}, 'sig2'))
        .rejects
        .toThrowError('Signature mismatch');
});

it('does not validate when no secret is set', async () => {
    const response = await getBlog({
        webhookSecret: undefined
    }).handleWebhookRequest({subdomain: 'test'}, 'sig2');
    expect(response.status).toBe(200);
    expect(await response.text()).toBe('ok');
});

it('clears template cache', async () => {
    const blog = getBlog();
    const response = await blog.handleWebhookRequest({
        subdomain: 'test',
        event: 'cache.templates'
    }, 'sig');
    expect(response.status).toBe(200);
    expect(await response.text()).toBe('ok');

    const cached = await blog.cacheService.getFromCache('LAST_TEMPLATE_CACHE_CLEARED_AT');
    expect(cached).toBeGreaterThan(0);
});

it('clears single cache', async () => {
    const blog = getBlog();

    await blog.cacheService.set('/test', {type: 'file'} as ResponseObject);

    const response = await blog.handleWebhookRequest({
        subdomain: 'test',
        event: 'cache.single',
        path: '/test'
    }, 'sig');
    expect(response.status).toBe(200);
    expect(await response.text()).toBe('ok');

    const cached = await blog.cacheService.getFromCache('/test');
    expect(cached).toBeUndefined();

});

it('clears all cache', async () => {
    const blog = getBlog();
    const response = await blog.handleWebhookRequest({
        subdomain: 'test',
        event: 'cache.all'
    }, 'sig');
    expect(response.status).toBe(200);
    expect(await response.text()).toBe('ok');

    const cached = await blog.cacheService.getFromCache('LAST_ALL_CACHE_CLEARED_AT');
    expect(cached).toBeGreaterThan(0);
});


it('validates signature', async () => {
    const blog = getBlog({
        webhookSecret: 'secret'
    });
    const data = {
        subdomain: 'test',
        event: 'cache.templates'
    };
    const response = await blog.handleWebhookRequest(
        data, 
        'e3b22f1e804080526e64fa2e757b575914560009bd0e034a664cc655cf57021e'
    );
    expect(response.status).toBe(200);
    expect(await response.text()).toBe('ok');
});

src/types.ts
export type CacheStore = {
    get(key: string): Promise<any>;
    set(key: string, value: any): Promise<any>;
    delete(key: string): Promise<any>;
}

export interface BlogOptions {

    /**
     * The subdomain given by Hyvor Blogs
     * Sign up at https://blogs.hyvor.com/console to create a new one
     * Get your subdomain from Console -> Settings -> Hosting
     */
    subdomain: string,

    /**
     * API Key to access the Delivery API of your blog
     * Console -> Settings -> API Keys -> Create an API key for Delivery API
     * Add the API key to your environment variables
    */
    deliveryApiKey: string,

    /**
     * Optional but recommended
     * As per the tutorial, you should set up a webhook with all cache events at
     * Console -> Settings -> Webhooks
     * Then, copy the secret here for webhook validation
     * If the secret is undefined, no validation is done
    */
    webhookSecret?: string,

    /**
     * Set up a custom cache store for caching
     * This must be an object with `get`, `set`, and `delete` methods (See `CacheStore` type)
     * [Keyv](https://keyvhq.js.org/) is a good option
     * If not provided, an in-memory store is used using Keyv
     */
    cache?: CacheStore 

    /**
     * This option is for package developers and testing purposes
     */
    hbApiEndpoint?: string,


    experimental?: {
        /**
         * If true, cache: 'no-store' is not sent in fetch requests
         * This is a workaround for a bug in Cloudflare Workers
         * https://github.com/cloudflare/workerd/issues/698
         */
        dontSetCacheInFetch?: boolean
    }

}

export interface FileResponse {
    type: 'file',
    at: number,
    cache: boolean,
    status: number,
    file_type: 'template' | 'asset' | 'media',
    content: string,
    mime_type: string,
    cache_control: string,
}

export interface RedirectResponse {
    type: 'redirect',
    at: number,
    cache: boolean,
    status: number,
    to: string
}

export type ResponseObject = FileResponse | RedirectResponse;

src/index.ts
import Keyv from "@keyvhq/core";
import { BlogOptions, CacheStore, ResponseObject } from "./types";
import { verifyHmacSha256 } from "./crypto";
 
export class Blog {

    public cacheService: CacheService;

    constructor(public options: BlogOptions) {
        this.cacheService = new CacheService(options);
    }

    public async handleBlogRequest(path: string) {

        path = path.startsWith('/') ? path : '/' + path;

        const apiEndpoint = 
            (this.options.hbApiEndpoint || 'https://blogs.hyvor.com/api/delivery/v0/') + 
            this.options.subdomain;
        
        const apiKey = this.options.deliveryApiKey;

        let responseObject = await this.cacheService.get(path);

        if (!responseObject) {

            const fetchOptions : RequestInit = {};

            if (this.options.experimental?.dontSetCacheInFetch !== true) {
                fetchOptions.cache = 'no-store';
            }

            const response = await fetch(`${apiEndpoint}?path=${path}&api_key=${apiKey}`, fetchOptions);

            if (!response.ok) {
                return new Response(null, {status: 500});
            }

            responseObject = await response.json() as ResponseObject;

            await this.cacheService.set(path, responseObject);
        }

        return this.responseObjectToResponse(responseObject);
    }

    /**
     * Returns 200 when failing on known errors to prevent webhook retries.
     */
    public async handleWebhookRequest(data: any, signature: string) {

        if (typeof data !== 'object' || data === null) {
            throw new Error('Invalid data type');
        }

        const subdomain = data.subdomain;

        if (subdomain !== this.options.subdomain) {
            throw new Error('Subdomain mismatch');
        }

        if (this.options.webhookSecret) {
            const contentString = JSON.stringify(data);

            const verify = await verifyHmacSha256(contentString, this.options.webhookSecret, signature); 

            if (!verify) {
                throw new Error('Signature mismatch');
            }

        }

        const event = data.event;

        if (event === 'cache.single') {
            await this.cacheService.clearSingleCache(data.path);
        } else if (event === 'cache.templates') {
            await this.cacheService.clearTemplateCache();
        } else if (event === 'cache.all') {
            await this.cacheService.clearAllCache();
        }

        return new Response('ok');

    }

    private responseObjectToResponse(responseObject: ResponseObject) {

        if (responseObject.type === 'redirect') {
            return Response.redirect(responseObject.to, responseObject.status);
        }

        const byteCharacters = atob(responseObject.content);

        // Convert the byte characters to an array buffer
        const byteNumbers = new Array(byteCharacters.length);
        for (let i = 0; i < byteCharacters.length; i++) {
            byteNumbers[i] = byteCharacters.charCodeAt(i);
        }
        const byteArray = new Uint8Array(byteNumbers);
        const arrayBuffer = byteArray.buffer;

        // Create a blob from the array buffer
        const blob = new Blob([arrayBuffer], { type: responseObject.mime_type });

        return new Response(blob, {
            status: responseObject.status,
            headers: {
                'Content-Type': responseObject.mime_type,
                'Cache-Control': responseObject.cache_control,
            }
        });


    }

}


export class CacheService {

    private readonly LAST_TEMPLATE_CACHE_CLEARED_AT = 'LAST_TEMPLATE_CACHE_CLEARED_AT';
    private readonly LAST_ALL_CACHE_CLEARED_AT = 'LAST_ALL_CACHE_CLEARED_AT';

    private store: CacheStore;

    constructor(private options: BlogOptions) {
        this.store = options.cache || new Keyv();
    }

    public async set(path: string, object: ResponseObject) {
        await this.store.set(this.getKey(path), object);
    }

    public async get(path: string) : Promise<null|ResponseObject> {

        const cached = await this.getFromCache(path);

        if (!cached)
            return null;

        if (typeof cached !== 'object')
            return null;

        const responseObject = cached as ResponseObject;
        const at = responseObject.at;

        const lastCacheClearedAt = await this.getFromCache(this.LAST_ALL_CACHE_CLEARED_AT) || 0;
        if (at < lastCacheClearedAt)
            return null;

        if (responseObject.type === 'file' && responseObject.file_type) {
            const lastTemplateCacheClearedAt = await this.getFromCache(this.LAST_TEMPLATE_CACHE_CLEARED_AT) || 0;
            if (at < lastTemplateCacheClearedAt)
                return null;
        }   

        return responseObject;
    }


    public async getFromCache(key: string) {
        return await this.store.get(this.getKey(key));
    }

    private getKey(key: string) {
        return `hyvor_blogs:${this.options.subdomain}:${key}`;
    }


    public async clearSingleCache(path: string) {
        await this.store.delete(this.getKey(path));
    }

    public async clearTemplateCache() {
        await this.store.set(
            this.getKey(this.LAST_TEMPLATE_CACHE_CLEARED_AT),
            Math.floor(Date.now() / 1000)    
        );
    }

    public async clearAllCache() {
        await this.store.set(
            this.getKey(this.LAST_ALL_CACHE_CLEARED_AT),
            Math.floor(Date.now() / 1000)    
        );
    }

}